### PR TITLE
[WIP] Make Ch06 notebook run with CUDA

### PR DESCRIPTION
**Pull request by Alfons0329**
* Linux
```$ uname -ar
<Machine_name> 4.15.0-65-generic 
74~16.04.1-Ubuntu SMP Wed Sep 18 09:51:44 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

```
* CUDA
```
$ /usr/local/cuda-9.2/bin/nvcc -V
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2018 NVIDIA Corporation
Built on Tue_Jun_12_23:07:04_CDT_2018
Cuda compilation tools, release 9.2, V9.2.148
```
* Pysyft
```
$ pip show syft                    
Name: syft
Version: 0.2.0a2
Summary: A Library for Private, Secure Deep Learning
Home-page: https://github.com/OpenMined/PySyft
Author: Andrew Trask
Author-email: contact@openmined.org
License: Apache-2.0
Location: /home/alfons0329/.local/lib/python3.6/site-packages
Requires: msgpack, torchvision, tblib, torch, Flask, websocket-client, flask-socketio, websockets, numpy, lz4, zstd
Required-by:
```
* Pytorch and torchvision
 --> This version is compatible with syft version 0.2.0a2

## 1. Runtime Error: {backend CUDA for arg...} in tutorial 6
* Problem be like
```
RuntimeErrorTraceback (most recent call last)
<timed exec> in <module>

~/.local/lib/python3.6/site-packages/torch/nn/modules/module.py in to(self, *args, **kwargs)
    424             return t.to(device, dtype if t.is_floating_point() else None, non_blocking)
    425 
--> 426         return self._apply(convert)
    427 
    428     def register_backward_hook(self, hook):

~/.local/lib/python3.6/site-packages/torch/nn/modules/module.py in _apply(self, fn)
    200     def _apply(self, fn):
    201         for module in self.children():
--> 202             module._apply(fn)
    203 
    204         def compute_should_use_set_data(tensor, tensor_applied):

~/.local/lib/python3.6/site-packages/torch/nn/modules/module.py in _apply(self, fn)
    225                 should_use_set_data = compute_should_use_set_data(param, param_applied)
    226                 if should_use_set_data:
--> 227                     param.data = param_applied
    228                 else:
    229                     assert isinstance(param, Parameter)

~/.local/lib/python3.6/site-packages/syft/frameworks/torch/hook/hook.py in data(self, new_data)
    330 
    331                 with torch.no_grad():
--> 332                     self.set_(new_data)
    333             return self
    334 

RuntimeError: Expected object of device type cuda but got device type cpu for argument #1 'self' in call to _th_set_
```
In the traning session

* Solution
Add `torch.set_default_tensor_type(torch.cuda.FloatTensor)`
in 

## (Appear after 1.) Runtime Error: {RuntimeError: Cannot re-initialize CUDA in forked subprocess.} in tutorial 6
* Problem be like
```
File "/home/alfons0329/.local/lib/python3.6/site-packages/syft/frameworks/torch/hook/hook.py", line 486, in new_tensor
    current_tensor = hook_self.torch.native_tensor(*args, **kwargs)
  File "/home/alfons0329/.local/lib/python3.6/site-packages/torch/cuda/__init__.py", line 191, in _lazy_init
    "Cannot re-initialize CUDA in forked subprocess. " + msg)
RuntimeError: Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method
```
* Solution
Force the spawn method in the beginning part near arguments

## Dec 3, 2019 Currently runnable

And it really works! With bigger batch size comses with bigger ram usage!!

Batch size == 8192

Batch size == 16384

### CODE REVIEW

**Comment by Alfons0329** [comment]
> Could you please check, whether we can get rid of
> 
> ```
> torch.set_default_tensor_type(torch.cuda.FloatTensor)
> ```
> 
> with the newest content of master?

No problem, but should I use
`pip install syft --upgrade`
Or `git pull`, I am afraid that the latter method will have some conflict since I have write some notes for my school study in it.
Thank you so much !

**Comment by midokura-silvia** [comment]
> No problem, but should I use
> `pip install syft --upgrade`
> Or `git pull`, I am afraid that the latter method will have some conflict since I have write some notes for my school study in it.
> Thank you so much !

If you have local commits on master you can do the following to have a clean state:
First create a new local branch to store you modifications:
`git checkout -b dev-alfons`
`git stash` or `git commit` (stash or commit your local changes (without commit) to the new branch)
Then get the newest state of master:
`git fetch`
`git checkout master`
`git reset --hard origin`
Then you will have a clean state of master. You can then either cherry-pick your commit or copy-paste the modified file.

**Comment by Alfons0329** [comment]
I am now at #2761, and the rest configuration has not been changed or reconfigured.
But the issue remains, with the png attached, thanks.

image

image

**Comment by midokura-silvia** [comment]
You are still running the pysyft code before the gpu commit. That's why this error appears.

**Comment by Alfons0329** [comment]
> You are still running the pysyft code before the gpu commit. That's why this error appears.

Yes I have done `git fetch`, `git checkout master` and `git reset --hard origin`
And my latest head is at 

```
 g reset --hard origin
HEAD is now at d8ede46 Add youtube link to README (#2761)
```
Please correct me if I have done something wrong since it is my first time open source contribution, thank you so much!

**Comment by gmuraru** [comment]
@Alfons0329 could you solve the conflicts and update the PR ? :)

**Comment by Alfons0329** [comment]
@gmuraru  Sorry for the late reply since I was traveling with my family during 28 Jan to 1 Feb.
I will try to fix it, or add some comments as soon as I finished my school research.
Thanks for the understanding.

**Comment by karlhigley** [comment]
@Alfons0329 It looks like the notebook tests are failing on CI, because the CI machines don't have GPUs. Can we make this change conditional on whether a GPU is available?

**Comment by Alfons0329** [comment]
Sorry my native language is not English, does "Can we make this change conditional on whether a GPU is available" means that: I can do a conditional version of the notebook for GPU (for example, set the argument to let the user define whether or not to use GPU)
Such as this pseudo code

```python
class args():
 self.gpu = true (Manually
 device = 'cuda' if cuda_is_available else 'cpu'
 if self.gpu = true:
 model = model.to(device)
else:
 # no op
```

Such that 2 versions of processing unit can be done for CI/CD

**Comment by karlhigley** [comment]
Ideally, we'd do the check to see if a GPU is available in the code, so the user doesn't have to specify, which seems like it should already by handled by `torch.cuda.is_available()` and the `use_cuda` variable.

**Comment by karlhigley** [comment]
Maybe we could do something like:
```
if use_cuda:
    torch.set_default_tensor_type(torch.cuda.FloatTensor)
    torch.multiprocessing.set_start_method('spawn', force=True)
```

**Comment by Alfons0329** [comment]
Okay, I will try to fix and figure out the rest tomorrow (Midnight here) and update what I have found or thoughts, attemps...etc. 
Thanks for the information!

**Comment by Alfons0329** [comment]
The unit test failed, seems to be the timeout in pytest.
I have found a timeout setting in test/notebook/test_notebooks.py and  it was set to 300 (at line 88, where the problem occurred)
(But the to change the setting to non-300 value, it only affect my forked repository, not the official to-be-merged repository) 
Not sure the next step to do.
Any suggestion helps!, Thanks a lot!

File: test/notebooks/test_notebooks.py [hunk 1]
```diff
@@ -100,7 +100,7 @@ def test_notebooks_basic(isolated_filesystem, notebook):
         notebook,
         "/dev/null",
         parameters={"epochs": 1, "n_test_batches": 5, "n_train_items": 64, "n_test_items": 64},
-        timeout=300,
+        timeout=1000,
```
Comment (karlhigley) [inline test/notebooks/test_notebooks.py:103]:
```suggestion
        timeout=300,
```

**Pull request closed without merging**